packages/serverpod/lib/src/server/serverpod.dart

```diff
@@ -911,49 +982,45 @@ class Serverpod {
   }
 
   /// Shuts down the Serverpod and all associated servers.
-  /// If [exitProcess] is set to false, the process will not exit at the end of the shutdown.
+  /// If [exitProcess] is set to false, the process will not exit at the end of
+  /// the shutdown.
   Future<void> shutdown({
     bool exitProcess = true,
     int? signalNumber,
   }) async {
     stdout.writeln(
         'SERVERPOD initiating shutdown, time: ${DateTime.now().toUtc()}');
 
-    var futures = [
-      _shutdownTestAuditor(),
-      _internalSession.close(),
-      redisController?.stop(),
-      server.shutdown(),
-      _webServer?.stop(),
-      _insightsServer?.shutdown(),
-      _futureCallManager?.stop(),
-      _healthCheckManager?.stop(),
-    ].nonNulls;
-
     Object? shutdownError;
-    await futures.wait.onError((ParallelWaitError e, stackTrace) {
-      shutdownError = e;
-      var errors = e.errors;
-      if (errors is Iterable<AsyncError?>) {
-        for (var error in errors.nonNulls) {
-          _reportException(
-            error.error,
-            error.stackTrace,
-            message: 'Error in server shutdown',
-          );
-        }
-      } else {
+
+    await _requestReceivingShutdownTasks.executeTasks(
+      onTaskError: (error, stack, id) {
```

Comment:
Nit: As it's all internal here (and there is really just the logger as a sensible implementation) I don't mind, but I find the "extra info as callbacks" pattern really unintuitive (there's also some in the legacy auth module).

Why not return a list of errors / overall status instead?